src/worker/ingestion/hooks.ts

```diff
@@ -0,0 +1,225 @@
+import { PluginEvent } from '@posthog/plugin-scaffold'
+import { captureException } from '@sentry/node'
+import { StatsD } from 'hot-shots'
+import fetch from 'node-fetch'
+import { format } from 'util'
+
+import { Action, Person } from '../../types'
+import { DB } from '../../utils/db/db'
+import { stringify } from '../../utils/utils'
+import { OrganizationManager } from './organization-manager'
+import { TeamManager } from './team-manager'
+
+export enum WebhookType {
+    Slack = 'slack',
+    Discord = 'discord',
+    Teams = 'teams',
+}
+
+export function determineWebhookType(url: string): WebhookType {
+    url = url.toLowerCase()
+    if (url.includes('slack.com')) {
+        return WebhookType.Slack
+    }
+    if (url.includes('discord.com')) {
+        return WebhookType.Discord
+    }
+    return WebhookType.Teams
+}
+
+export function getUserDetails(
+    event: PluginEvent,
+    person: Person | undefined,
+    siteUrl: string,
+    webhookType: WebhookType
+): [string, string] {
+    if (!person) {
+        return ['undefined', 'undefined']
+    }
+    const userName = stringify(person.properties?.['email'] || event.distinct_id)
+    let userMarkdown: string
+    if (webhookType === WebhookType.Slack) {
+        userMarkdown = `<${siteUrl}/person/${event.distinct_id}|${userName}>`
+    } else {
+        userMarkdown = `[${userName}](${siteUrl}/person/${event.distinct_id})`
+    }
+    return [userName, userMarkdown]
+}
+
+export function getActionDetails(action: Action, siteUrl: string, webhookType: WebhookType): [string, string] {
+    const actionName = stringify(action.name)
+    let actionMarkdown: string
+    if (webhookType === WebhookType.Slack) {
+        actionMarkdown = `<${siteUrl}/action/${action.id}|${action.name}>`
+    } else {
+        actionMarkdown = `[${action.name}](${siteUrl}/action/${action.id})`
+    }
+    return [actionName, actionMarkdown]
+}
+
+export function getTokens(messageFormat: string): [string[], string] {
+    const matchedTokens = messageFormat.match(/(?<=\[)(.*?)(?=\])/g) || []
+    let tokenizedMessage = messageFormat
+    if (matchedTokens.length) {
+        tokenizedMessage = tokenizedMessage.replace(/\[(.*?)\]/g, '%s')
+    }
+    return [matchedTokens, tokenizedMessage]
+}
+
+export function getValueOfToken(
+    action: Action,
+    event: PluginEvent,
+    person: Person | undefined,
+    siteUrl: string,
+    webhookType: WebhookType,
+    tokenParts: string[]
+): [string, string] {
+    let text = ''
+    let markdown = ''
+
+    if (tokenParts[0] === 'user') {
+        if (tokenParts[1] == 'name') {
```

Comment:
```suggestion
        if (tokenParts[1] === 'name') {
```

and a few below, too! Python->JS conversion can get so annoying 😂